feat: Virtual DOM 비교와 이벤트 처리 구현하기

nsj9164 opened the pull request:
## 📝 작업 내용
- Virtual DOM 비교(diffing) 알고리즘 구현
- 이벤트 처리 및 Synthetic Event 시스템 구현
- 추가적인 훅(useEffect) 구현 및 컴포넌트 생명주기 관리

[Todo앱 결과물]
![image](https://github.com/user-attachments/assets/baa993d1-8de6-4c6c-b436-30e9d2e4656e)

nsj9164 committed:
feat: Virtual DOM diffing, 이벤트 처리 시스템, useEffect 훅 구현

nsj9164 committed:
fix: diff에 key 기반 자식 비교 로직 추가

nsj9164 committed:
fix: form과 onSubmit handler 추가

eastroots92 commented on src/libs/diff.js:
```diff
@@ -0,0 +1,144 @@
+
+  updateProps(parentRoot, oldVNode.props, newVNode.props);
+
+  const oldChildren = Array.isArray(oldVNode.props?.children)
+    ? oldVNode.props.children
+    : [oldVNode.props?.children].filter(Boolean);
```
P4; 질문.

children이 array가 아닌 경우 boolean으로 필터를 해주는 이유는 무엇인가요?
array인 경우에는 따로 boolean으로 필터를 안해줘도 되는 것인지도 궁금해요.

eastroots92 commented on src/libs/diff.js:
```diff
@@ -0,0 +1,144 @@
+
+  oldChildren.forEach((child, i) => {
+    const key = child?.props?.key;
+    if (key != null) oldKeyed[key] = { child, index: i };
```
p5;

가능하다면 얕은 비교보단 깊은 비교를 하는 것을 권장드려요.

얕은 비교와 깊은 비교 이 둘의 차이에 대해 혹시 모르신다면 쓰윽 찾아보셔도 좋을 것 같아요.

eastroots92 commented on src/libs/diff.js:
```diff
@@ -0,0 +1,144 @@
+
+  Object.keys(newProps).forEach((key) => {
+    if (key === "children") return;
+    if (oldProps[key] !== newProps[key]) {
```
현재 해당 하위 if else 문의 복잡도가 매우 높은 것 같아요.

조금 더 간결하게 바꿔 볼 수 없을까요?

eastroots92 commented on src/libs/diff.js:
```diff
@@ -0,0 +1,144 @@
+        }
+
+        // 기존 handler 제거 후 새로 바인딩 (이벤트 handler 최신으로 교체)
+        if (dom._handlers?.[eventType]) {
```
P3;
이전 이벤트와 달라진 것이 없다면 굳이 제거하지 않아도 좋을 것 같은데 어떠신가요!?

eastroots92 commented on src/libs/events.js:
```diff
@@ -0,0 +1,25 @@
+export function handleEvent(event) {
```
P3;

위 diff에 해당 함수를 적용하려 했던 것으로 보여지는데요.
현재는 해당 함수가 적용되어있지 않은 것 같아요.

이 처럼 diff 코드의 맥락을 나눠서 적용해보면 좋을 것 같아요.

추가 질문으로. 이렇게 코드를 나눴을 때 어떤 장점을 가지는지 스윽 생각해보시면 좋을 것 같아요!

eastroots92 commented on src/libs/render.js:
```diff
@@ -1,38 +1,78 @@
+let oldVNode = null;
+let rootContainer = null;
+
 export function render(vdom, container) {
```
P3;

현재 render와 diff내 update 해주는 함수와 코드 중복이 있는 것 같아요.
이 부분을 효율적으로 처리 할 수 있을 것 같은데 개선해볼 수 있을까요?

eastroots92 commented on src/libs/useEffect.js:
```diff
@@ -0,0 +1,31 @@
+const effectStack = [];
+
+export function useEffect(callback, dependencies) {
```
useEffect와 useState의 실행 시점은 각기 다른데요.
보통 useState 반영이 끝난 후 useEffect가 실행되어요.

이 순서를 보장해야 할 것 같은데 어떻게 보장 할 수 있을까요?

eastroots92 reviewed (commented):
안녕하세요 하나님!

고민해보면 좋을 내용에 대해 리뷰 남겨 두었어요.

3주차 과제가 고민 할 것도, 공부할 것도 많아 쉽지 않았을 텐데 포기하지 않고 쭈욱 하시는 모습이 정말 대단하시고 멋지다고 생각해요.
정말 고생 많으셨어요.

nsj9164 committed:
fix : 코드리뷰 반영하여 diff 함수 개선_#1

nsj9164 committed:
fix : 코드리뷰 반영하여 diff 함수 개선_#2

nsj9164 committed:
fix : 코드리뷰 반영 (#3) - useEffect 함수 개선
- 상태 반영 이후 useEffect 실행을 보장하도록 수정
- cleanup 함수가 정확히 실행되도록 수정

nsj9164 committed:
fix : 코드리뷰 반영하여 함수 개선_#4

- props 처리 로직을 applyProps로 통합하여 중복 제거
- normalizeChildren에서 false, null, undefined 필터링 추가
- createDom에서 boolean 값에 대한 방어 코드 추가
- useEffect에서 cleanup과 callback 실행 조건 명확화
- checkbox의 checked 속성이 반영되지 않던 문제 수정
